### PR TITLE
[D][Security] Enforce typed dangerous-action blast-radius caps

### LINKED ISSUE #311: [D][Security] Enforce typed dangerous-action blast-radius caps
Part of #170 (Epic D, Phase 2).

## Goal

Evaluate typed dangerous-action limits before execution and deny calls that exceed per-action or delegation-lineage blast-radius caps. The signed receipt must explain the bounded reason without leaking the sensitive arguments it protected.

## Verified current gap

Ardur already performs useful pre-action checks for tool allowlists, resource scope/cwd, side-effect classes, per-class call counts, and closed-world AAT/DRP argument constraints. However:

- there is no typed policy vocabulary for destructive target cardinality, bytes/objects affected, secret sensitivity, or destination risk;
- lineage counters conserve calls, not destructive impact;
- DRP `boundaries` strings are signed commitments but are not an executable evaluator;
- tool-name heuristics alone cannot prove the semantic blast radius of arbitrary tools.

## What to build

- Define a versioned, declarative risk-fact contract produced from authenticated tool schemas plus normalized invocation arguments, not prompt text.
- Cover at minimum: destructive target count, bytes/objects affected where knowable, secret/sensitive-data access class, network destination class, filesystem scope, and operation irreversibility.
- Define per-action caps and atomically conserved per-session/per-agent/per-lineage caps. Delegation may only preserve or narrow every cap.
- Require each governed dangerous tool to declare which risk facts are mandatory. Missing, ambiguous, malformed, overflowed, or unknown mandatory facts must deny before execution.
- Keep tool-specific extractors isolated and testable. Do not execute the action or perform an unbounded discovery scan to estimate its blast radius.
- Reuse closed-world argument constraints where they are sufficient, but do not pretend a range constraint is an aggregate lineage ledger.
- Emit stable signed denial reasons and bounded coarse measurements/digests only. Do not record secret values, raw paths, command bodies, recipient lists, or high-cardinality target sets.
- Export low-cardinality metrics for permits/denials/reservations by risk class and reason.

## Acceptance criteria

- [ ] Bulk delete above the configured target cap is denied before executor invocation.
- [ ] Secret-class access and out-of-policy network/filesystem destinations deny pre-action with stable reasons.
- [ ] Concurrent siblings cannot exceed a lineage-wide destructive-impact ceiling.
- [ ] Missing/unknown risk facts fail closed for tools that declare them mandatory.
- [ ] Risk caps attenuate transitively across A→B→C and widening fixtures fail.
- [ ] Retry/idempotency, reservation release/settlement, overflow, normalization, and crash cases are tested.
- [ ] Signed receipts remain privacy-bounded and offline-verifiable.
- [ ] Existing safe/resource-free tools remain usable without inventing irrelevant risk grants.
- [ ] Unit, adversarial, concurrency, integration, privacy, and relevant full-suite tests pass.
- [ ] PR targets `dev` from an isolated EXTENDED worktree with signed DCO commits and green CI.

## Security and cost boundary

Preflight estimators can be conservative and cause false denials; silent underestimation is worse. Any external inventory query used for estimation has latency and request-cost implications and must be bounded, authenticated, separately authorized, and excluded from the first implementation unless explicitly justified.

## Primary references

- AAT draft current chain/argument enforcement model: https://datatracker.ietf.org/doc/draft-niyikiza-oauth-attenuating-agent-tokens/
- DRP draft-10 pre-action boundaries and sub-receipts: https://datatracker.ietf.org/doc/html/draft-nelson-agent-delegation-receipts-10

### PR DESCRIPTION
**Pull request by gnanirahulnutakki**
## Summary

Closes #311.

Adds an optional signed and attenuable `risk_budget` authority for dangerous tool actions. Trusted, JSON-Schema-bound tool contracts extract typed blast-radius facts before execution; one durable transaction reserves numeric impact across session, agent, and delegation-lineage scopes before ordinary policy can permit the action.

## What changed

- define immutable, digest-bound `ToolRiskContract` schemas with a closed extractor vocabulary;
- support numeric facts for destructive targets, objects affected, and bytes affected;
- support closed categorical facts for secret sensitivity, destination risk, filesystem scope, and irreversibility;
- normalize and attenuate optional signed passport `risk_budget` claims with subset-only delegation;
- atomically reserve all numeric facts across session/agent/lineage scopes with `flock`, `fsync`, and same-filesystem atomic publication;
- require explicit executor outcomes: commit successful/started work, release only work proven not started, and quarantine uncertain crash residue without refund;
- preserve replay tombstones and retry-safe lifecycle receipt outbox state;
- integrate preflight/outcome handling into the proxy and HTTP `/risk/outcome` boundary;
- emit privacy-safe receipts and bounded metrics without paths, destinations, request IDs, account data, or secret values;
- add ADR-026, operator/reference documentation, DRP mapping/inventory updates, and synchronized site mirrors.

## Security and reliability properties

- Prompt text, caller-provided labels, and unauthenticated MCP annotations cannot lower risk.
- Unknown mandatory facts, contract digest mismatch, malformed facts, and current DRP projection all fail closed.
- Reservation is durable before permit; crash, receipt failure, or missing outcome cannot silently restore destructive authority.
- Cached sessions retain the originally persisted risk-policy snapshot across policy rotation.
- Receipt append-once recovery uses thread and inter-process locking and reuses the persisted fact digest.
- Existing passports without `risk_budget` preserve current behavior.

## Validation

- `44 passed` — focused risk-budget suite, including multiprocessing deduplication, crash recovery, policy rotation/cache refresh, immutable contracts, replay tombstones, and quarantine non-refund.
- `291 passed in 9.03s` — risk budget, HTTP, passport, delegation, lineage, proxy, receipt, DRP conformance, and mapping tests on the exact post-formatter commit.
- Supplemental isolated Python 3.12.8 suite: `1855 passed, 34 skipped, 6 warnings in 87.16s`.
- Ruff, formatter, JSON, private-key, hardcoded-secret, line-ending, and file-size hooks passed.
- Documentation/site validation: 117 source pages, 113 mirrored artifacts, 287 rendered Hugo pages, 116 static files, 10 public claims, and clean rendered links.
- Final CodeRabbit review covered all 31 files with no critical or warning #311 findings.

## Review disposition

The final independent review found one minor pre-existing gap: the CLI API-token trim test bypasses CLI argument loading. It is not caused by this branch and is tracked separately in #315 with focused security/test acceptance criteria.

## Cost and operability

Risk extraction is local, bounded, and deterministic: it performs no inventory scan or network request, avoiding external request charges, data egress, and runtime availability coupling. Ledger/tombstone retention is bounded, but operators should still monitor the private state directory and receipt storage for volume growth.

## Standards references

- AAT draft `-01`: https://datatracker.ietf.org/doc/draft-niyikiza-oauth-attenuating-agent-tokens/
- DRP draft `-10`: https://datatracker.ietf.org/doc/html/draft-nelson-agent-delegation-receipts-10
- MCP tools 2025-11-25: https://modelcontextprotocol.io/specification/2025-11-25/server/tools
- RFC 8785: https://www.rfc-editor.org/rfc/rfc8785.html

## Promotion boundary

This PR targets `dev`. Human review is mandatory. It does not modify or promote to `main`.

## Summary by CodeRabbit

- **New Features**
  - Added optional typed `risk_budget` support to Mission Passports, including delegated attenuation and stricter validation behavior.
  - Introduced a file-backed risk-budget ledger with atomic reservations, explicit committed vs released lifecycle outcomes, replay/quarantine handling, and privacy-bounded receipts.
  - Added risk-budget metrics plus governance proxy API support for recording risk outcomes (including `risk_request_id` and a `/risk/outcome` flow).
- **Bug Fixes**
  - Hardened fail-closed enforcement for invalid/unknown risk states and improved denial/outcome mapping and bearer-auth edge cases.
- **Documentation**
  - Published ADR-026 and updated reference/spec pages and public-facing README sections.
- **Tests**
  - Added extensive lifecycle, delegation, and end-to-end HTTP coverage for contracts, reservations, replay, and quarantine semantics.

### CODE REVIEW

File: python/tests/test_risk_budget.py [hunk 1]
```diff
@@ -0,0 +1,1385 @@
...
+        try:
+            barrier.wait()
+            result = _reserve(
+                ledgers[index],
+                request_id=f"request-{index}",
+                objects=1,
+                ceiling=1,
+            )
+            results.append(result.accepted)
+        except BaseException as exc:  # pragma: no cover - asserted below
```
[the remark this answers is not in the document]
Comment (gnanirahulnutakki) [reply python/tests/test_risk_budget.py]: Fixed in signed DCO commit b34713f4ae68a76b901ca9a918742de35bd7513f. The concurrency test worker now catches Exception rather than BaseException, so process-control exceptions are not swallowed. Ruff format/lint passed and python/tests/test_risk_budget.py is green: 44 passed in 0.98s.

**Comment by gnanirahulnutakki** [comment]
Review-closure update on exact head `4a2f3160cb63b9e24630bdf12978d3336d43de65`:

- Fixed the valid #311 findings: complete issuer-owned claim protection; canonical UUID overrides; Mission Declaration and mission-ref risk-policy preservation; retained-tool policy projection; lineage-namespaced request identity; global cross-lineage agent accounting; copy-then-persist pruning; receipt append under session-chain coordination; and pending-lifecycle finalization guards.
- Decomposed the independent fail-open TLS startup seam as #317 instead of expanding this focused PR.
- No migration is included for the prior per-lineage draft ledger because #311 and that storage shape have never merged or shipped. The final global ledger makes agent ceilings atomic across lineages; ADR-026 records the serialization/throughput trade-off.
- Local CodeRabbit final review: zero findings.
- Exact final local gates: 52 focused risk tests; 373 compatibility tests; supplemental full Python suite 1,863 passed / 34 skipped / 6 known warnings; all repository hooks green; 117 source pages / 113 mirrors / 10 public claims; Hugo 287 pages / 116 static files.
- Commit signature is good ED25519 and DCO is present. The owned branch was pushed by normal fast-forward; `dev` and `main` are untouched.

Fresh CI and CodeRabbit on `4a2f316` are still required. Earlier green checks on `b34713f4` are historical only.
